routers/repo/issue.go

```diff
@@ -1237,6 +1250,120 @@ func UpdateIssueAssignee(ctx *context.Context) {
 	})
 }
 
+func isLegalReviewRequest(reviewer, doer *models.User, isAdd bool, issue *models.Issue) error {
+	if reviewer.IsOrganization() {
+		return fmt.Errorf("Organization can't be added as reviewer [user_id: %d, repo_id: %d]", reviewer.ID, issue.PullRequest.BaseRepo.ID)
+	}
+	if doer.IsOrganization() {
+		return fmt.Errorf("Organization can't be doer to add reviewer [user_id: %d, repo_id: %d]", doer.ID, issue.PullRequest.BaseRepo.ID)
+	}
+
+	permReviewer, err := models.GetUserRepoPermission(issue.Repo, reviewer)
+	if err != nil {
+		return err
+	}
+
+	permDoer, err := models.GetUserRepoPermission(issue.Repo, doer)
+	if err != nil {
+		return err
+	}
+
+	lastreview, err := models.GetReviewerByIssueIDAndUserID(issue.ID, reviewer.ID)
+	if err != nil {
+		return err
+	}
+
+	var pemResult bool
+	if isAdd {
+		pemResult = permReviewer.CanAccessAny(models.AccessModeRead, models.UnitTypeCode, models.UnitTypeIssues, models.UnitTypePullRequests)
+		if !pemResult {
+			return fmt.Errorf("Reviewer can't read [user_id: %d, repo_name: %s]", reviewer.ID, issue.Repo.Name)
+		}
+
+		if doer.ID == issue.PosterID && lastreview != nil && lastreview.Type != models.ReviewTypeRequest {
+			return nil
+		}
+
+		pemResult = permDoer.CanAccessAny(models.AccessModeWrite, models.UnitTypeCode, models.UnitTypeIssues, models.UnitTypePullRequests)
+		if !pemResult {
+			return fmt.Errorf("Doer can't write [user_id: %d, repo_name: %s]", doer.ID, issue.Repo.Name)
+		}
+
+		if doer.ID == reviewer.ID {
+			return fmt.Errorf("doer can't be reviewer [user_id: %d, repo_name: %s]", doer.ID, issue.Repo.Name)
+		}
+
+		if reviewer.ID == issue.PosterID {
+			return fmt.Errorf("poster of pr can't be reviewer [user_id: %d, repo_name: %s]", reviewer.ID, issue.Repo.Name)
+		}
+	} else {
+		if lastreview.Type == models.ReviewTypeRequest && lastreview.Content == doer.Name {
```

Comment:
The requester shouldn't be able to revoke the request unless it's an admin. Otherwise the requester can nag the reviewers by adding/removing them repeatedly.